core/async/src/test_loop/delay_sender.rs

```diff
@@ -0,0 +1,69 @@
+use std::{sync::Arc, time::Duration};
+
+use crate::messaging;
+
+use super::multi_instance::IndexedDelaySender;
+
+/// Interface to send an event with a delay (in virtual time). It can be
+/// converted to a Sender for any message type that can be converted into
+/// the event type, so that a DelaySender given by the test loop may be passed
+/// to production code that expects a Sender.
+pub struct DelaySender<Event> {
```

Comment:
is there any specific reason why you don't want every Sender to be DelaySender? Like the delay may be implemented on top of any sender anyway, right? Also, do you want to allow each DelaySender to implement the delay differently?

Reply:
Sender is a production type. DelaySender is a type specific to the TestLoop framework, it isn't used for production. (I suppose there's a case to be made that in production we could also do this, to implement periodic timers, so that even timer logic can be shared between production and tests, but the benefit isn't a clear yet.)

The DelaySender is not implemented as a Sender plus a delay. It's the other way around. A Sender (that is provided by the TestLoop to send into the loop's event queue) is implemented by redirecting messages to the DelaySender with a delay of zero.

The DelaySender actually only has three implementations:
 - Routes any (delayed) messages to the TestLoop (the most important one)
 - Relays a message to another DelaySender by calling .into() on the message type to convert it to a wider type
 - Adds an index to the message relays it to another DelaySender that takes a (usize, Event).

... which, necessitates an abstraction.